cli/module_registry.go

```diff
@@ -380,6 +399,67 @@ func getNextModuleUploadRequest(file *os.File) (*apppb.UploadModuleFileRequest,
 	}, nil
 }
 
+func validateModuleFile(client *viamClient, moduleID moduleID, tarballPath, version string) error {
+	getModuleResp, err := client.getModule(moduleID)
+	if err != nil {
+		return err
+	}
+	entrypoint, err := getEntrypointForVersion(getModuleResp.Module, version)
+	if err != nil {
+		return err
+	}
+	//nolint:gosec
+	file, err := os.Open(tarballPath)
+	if err != nil {
+		return err
+	}
+	// TODO(APP-2226): support .tar.xz
+	if !strings.HasSuffix(file.Name(), ".tar.gz") {
+		return errors.New("you must upload your module in the form of a .tar.gz")
+	}
+	archive, err := gzip.NewReader(file)
+	if err != nil {
+		return err
+	}
+
+	tarReader := tar.NewReader(archive)
+	filesWithSameNameAsEntrypoint := []string{}
+	for {
+		if err := client.c.Context.Err(); err != nil {
+			return err
+		}
+		header, err := tarReader.Next()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return errors.Wrapf(err, "error reading %s", file.Name())
+		}
+		path := header.Name
+
+		// if path == entrypoint, we have found the right file
+		if filepath.Clean(path) == filepath.Clean(entrypoint) {
+			info := header.FileInfo()
+			if info.Mode().Perm()&0o100 == 0 {
+				return errors.Errorf(
+					"the provided tarball %q contained a file at the entrypoint %q, but that file is not marked as executable",
+					tarballPath, entrypoint)
+			}
+			// executable file at entrypoint. validation succeeded.
+			return nil
+		}
+		if filepath.Base(path) == filepath.Base(entrypoint) {
+			filesWithSameNameAsEntrypoint = append(filesWithSameNameAsEntrypoint, path)
+		}
+	}
+	extraErrInfo := ""
+	if len(filesWithSameNameAsEntrypoint) > 0 {
+		extraErrInfo = fmt.Sprintf(". did you mean to set your entrypoint to %v?", filesWithSameNameAsEntrypoint)
```

Comment:
Showing the list of files when len > 1 might look more confusing than helpful. Maybe show only the first item?

Reply:
IMO i kind of think that showing the whole list is better. Would mean that the next time you went to upload if you had multiple of the same file you probs won't have the error any more if you removed all the files.

Reply:
Agree with you @michaellee1019 that it isn't a great msg if len > 1 but also not sure that it is better to just show the first item. But also agree with @RoxyFarhad. Planning to move forward with the current msg but happy to fix this myself if it becomes a problem in the future and confuses users.